File: ai-service/routers/health.py
"""Health check endpoints for service monitoring and readiness probes."""

import logging
from typing import Any

from fastapi import APIRouter
from fastapi.responses import JSONResponse

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/health", tags=["Health"])


@router.get("")
async def health_check() -> dict[str, str]:
    """Return basic service liveness status."""
    return {"status": "ok"}


@router.get("/ready")
async def readiness_check() -> JSONResponse:
    """Return service readiness including dependency connectivity."""
    from services import minio as minio_service
    from services import neo4j as neo4j_service
    from services import postgres as postgres_service
    from services import redis as redis_service
    from services import snowflake as snowflake_service

    checks: dict[str, str] = {}

    try:
        pool = postgres_service._pool
        if pool:
            pg_ok = await postgres_service.health_check(pool)
            checks["postgresql"] = "ok" if pg_ok else "error"
        else:
            checks["postgresql"] = "not initialized"
    except Exception as e:
        checks["postgresql"] = f"error: {e}"

    try:
        driver = neo4j_service._driver
        if driver:
            neo4j_ok = await neo4j_service.health_check(driver)
            checks["neo4j"] = "ok" if neo4j_ok else "error"
        else:
            checks["neo4j"] = "not initialized"
    except Exception as e:
        checks["neo4j"] = f"error: {e}"

    try:
        client = redis_service._client
        if client:
            redis_ok = await redis_service.health_check(client)
            checks["redis"] = "ok" if redis_ok else "error"
        else:
            checks["redis"] = "not initialized"
    except Exception as e:
        checks["redis"] = f"error: {e}"

    try:
        minio_client = minio_service._client
        if minio_client:
            minio_ok = minio_service.health_check(minio_client)
            checks["minio"] = "ok" if minio_ok else "error"
        else:
            checks["minio"] = "not initialized"
    except Exception as e:
        checks["minio"] = f"error: {e}"

    try:
        sf_ok = await snowflake_service.health_check()
        checks["snowflake"] = "ok" if sf_ok else "error"
    except Exception as e:
        checks["snowflake"] = f"error: {e}"

    all_ok = all(v == "ok" for v in checks.values())
    status_code = 200 if all_ok else 503

    return JSONResponse(
        status_code=status_code,
        content={
            "status": "ready" if all_ok else "degraded",
            "checks": checks,
        },
    )

File: backend/src/routes/health.ts
import type { FastifyInstance } from 'fastify';

import { postgresService } from '../services/postgresService.js';
import { neo4jService } from '../services/neo4jService.js';
import { redisService } from '../services/redisService.js';
import { minioService } from '../services/minioService.js';
import { snowflakeService } from '../services/snowflakeService.js';

type ServiceStatus = 'ok' | 'down';

interface ServiceHealthMap {
  postgresql: ServiceStatus;
  neo4j: ServiceStatus;
  redis: ServiceStatus;
  minio: ServiceStatus;
  snowflake: ServiceStatus;
}

async function checkAllServices(): Promise<ServiceHealthMap> {
  const [pg, neo4j, redis, minio, sf] = await Promise.all([
    postgresService.healthCheck(),
    neo4jService.healthCheck(),
    redisService.healthCheck(),
    minioService.healthCheck(),
    snowflakeService.healthCheck(),
  ]);

  return {
    postgresql: pg ? 'ok' : 'down',
    neo4j: neo4j ? 'ok' : 'down',
    redis: redis ? 'ok' : 'down',
    minio: minio ? 'ok' : 'down',
    snowflake: sf ? 'ok' : 'down',
  };
}

export async function healthRoutes(app: FastifyInstance): Promise<void> {
  /**
   * GET /health
   * Returns service status with individual dependency health.
   */
  app.get('/', async (_request, reply) => {
    const services = await checkAllServices();

    return reply.send({
      status: 'ok',
      version: '0.1.0',
      timestamp: new Date().toISOString(),
      services,
    });
  });

  /**
   * GET /health/ready
   * Readiness probe. Returns 200 if all services are healthy, 503 otherwise.
   */
  app.get('/ready', async (_request, reply) => {
    const services = await checkAllServices();
    const allHealthy = Object.values(services).every((s) => s === 'ok');

    const statusCode = allHealthy ? 200 : 503;

    return reply.status(statusCode).send({
      status: allHealthy ? 'ready' : 'not_ready',
      timestamp: new Date().toISOString(),
      services,
    });
  });
}
